### PR TITLE
pass options to jwt.verify to be able to verify audience and issuer. Also fix example.

### PR DESCRIPTION
**Pull request by solepano**
This PR adds:

1. Ability to pass options for verifying JWT:
```javascript
server.auth.strategy('token', 'jwt', {
    key: privateKey,
    validateFunc: validate,
    audience: 'http://myapi/protected',
    issuer: 'http://issuer',
    algorithms: ['RS256'],
    subject: 'myRequiredSubject'
});
```   
2. Joi validation for strategy options
3. Enhanced error messages in case of failed validation

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -62,9 +62,8 @@ internals.implementation = function (server, options) {
 
       getKey(request, token, function(err, key, extraInfo){
         if (err) { return reply(Boom.wrap(err)); }
-
         // handle err
-        jwt.verify(token, key, function(err, decoded) {
+        jwt.verify(token, key, settings, function(err, decoded) {
```
Comment (dschenkelman) [inline lib/index.js]: considering we are adding two more settings, could you please add some validation for the `options` object using `Joi`?

We should have done this before, but this is a good opportunity to get some validation and tests for it.

Thanks,

File: test/index.js [hunk 2]
```diff
@@ -313,4 +314,105 @@ describe('Token', function () {
...
+        { method: 'POST', path: '/token', handler: tokenHandler, config: { auth: 'default' } }
+      ]);
+    });
+    
+    it('fails if token audience is empty', function (done) {
+  
+      var request = { method: 'POST', url: '/token', headers: { authorization: tokenHeader('john') } };
+
+      server.inject(request, function (res) {
+        expect(res.result.message).to.equal('Invalid signature received for JSON Web Token validation');
```
Comment (dschenkelman) [inline test/index.js]: this error message looks weird. why is this related to the signature?

File: test/index.js [hunk 3]
```diff
@@ -313,4 +314,105 @@ describe('Token', function () {
...
+        { method: 'POST', path: '/token', handler: tokenHandler, config: { auth: 'default' } }
+      ]);
+    });
+    
+    it('fails if token issuer is empty', function (done) {
+  
+      var request = { method: 'POST', url: '/token', headers: { authorization: tokenHeader('john') } };
+
+      server.inject(request, function (res) {
+        expect(res.result.message).to.equal('Invalid signature received for JSON Web Token validation');
```
Comment (dschenkelman) [inline test/index.js]: same here, if the problem is the issuer, why does it talk about the signature?

Comment (dschenkelman) [inline test/index.js]: this is the right message for this case

File: README.md [hunk 4]
```diff
@@ -90,3 +94,16 @@ server.register(require('hapi-auth-jwt'), function (error) {
...
 
 ```
+
+You can specify audience, issuer, algorithms and/or subject as well:
+
+```javascript
+server.auth.strategy('token', 'jwt', {
+    key: privateKey,
+    validateFunc: validate,
+    audience: 'http://myapi/protected',
```
Comment (dschenkelman) [inline README.md]: should we put this in a `tokenOptions` object? Not sure. Bouncing ideas

File: lib/index.js [hunk 5]
```diff
@@ -3,10 +3,17 @@
 var Boom = require('boom');
 var Hoek = require('hoek');
 var jwt  = require('jsonwebtoken');
+var Joi = require('joi');
 
+var optionsSchema = Joi.object().keys({
+  key: Joi.alternatives().try(Joi.string(),Joi.func()).required(),
```
Comment (dschenkelman) [inline lib/index.js]: this is excellent

File: lib/index.js [hunk 6]
```diff
@@ -3,10 +3,17 @@
 var Boom = require('boom');
 var Hoek = require('hoek');
 var jwt  = require('jsonwebtoken');
+var Joi = require('joi');
 
+var optionsSchema = Joi.object().keys({
```
Comment (dschenkelman) [inline lib/index.js]: this is excellent

File: test/index.js [hunk 7]
```diff
@@ -161,7 +162,7 @@ describe('Token', function () {
     var request = { method: 'POST', url: '/token', headers: { authorization: tokenHeader('john', { expiresIn: -10 }) } };
 
     server.inject(request, function (res) {
-      expect(res.result.message).to.equal('Expired token received for JSON Web Token validation');
+      expect(res.result.message).to.equal('jwt expired');
```
Comment (dschenkelman) [inline test/index.js]: this message is too short. think that will be descriptive enough? same for next one

File: test/index.js [hunk 8]
```diff
@@ -313,4 +314,434 @@ describe('Token', function () {
     done();
   });
 
+  describe('when audience is specified for validation', function(){
```
Comment (dschenkelman) [inline test/index.js]: please, update to `'when a single audience is...'` to clearly diff from array case. Since this one shows up first, I was looking for the array tests in this describe and they are in another one (which is good)

File: README.md (hunk 4 above)
Comment (solepano) [inline README.md]: could be an option, but I think it's fine as it is

File: test/index.js (hunk 7 above)
Comment (solepano) [inline test/index.js]: it is the error message from jsonwebtoken. It's not TOO descriptive but...

Do you want to wrap all errors in a more descriptive error? don't know if its worth it

File: README.md (hunk 4 above)
Comment (solepano) [inline README.md]: having all options in same obj gives same experience as express-jwt:
```
var jwt = require('express-jwt');
var jwtCheck = jwt({
    secret: jwks.expressJwtSecret({
        cache: true,
        rateLimit: true,
        jwksRequestsPerMinute: 5,
        jwksUri: "https://solepano.auth0.com/.well-known/jwks.json"
    }),
    audience: 'https://auth0.auth0.com/api/admin',
    issuer: "https://solepano.auth0.com/",
    algorithms: ['RS256']
});
```

Comment (solepano) [inline README.md]: which is the objective of this PR

**Comment by solepano** [comment]
this is was tested and is ready to be merged @dschenkelman
